File: algorithm/LeetCode/Balanced Binary Tree/Balanced Binary Tree.java
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
public class Solution {
    public boolean isBalanced(TreeNode root) {
        if(root == null)
            return true; 
        if(root.left == null && root.right == null)
            return true;
        if(Math.abs(getDep(root.left) - getDep(root.right)) > 1)
            return false;
        return isBalanced(root.left) && isBalanced(root.right);
    }
    
    
    public int getDep(TreeNode node){
        if(node == null)
            return 0;
        else
            return 1 + Math.max(getDep(node.left), getDep(node.right));
    }
}

File: algorithm/LeetCode/Balanced Binary Tree/Balanced Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode* root) {
        int dep;
        checkBalance(root, dep);
    }
    bool checkBalance(TreeNode * root, int &dep)
    {
        if(root == NULL){
            dep = 0;
            return true;
        }
        int leftDep, rightDep;
        bool isLeftBal = checkBalance(root->left, leftDep);
        bool isRightBal = checkBalance(root->right, rightDep);

        dep = max(leftDep, rightDep) + 1;
        return isLeftBal && isRightBal && (abs(leftDep - rightDep) <= 1);
    }
};